components/org.wso2.carbon.identity.password.expiry/src/main/java/org/wso2/carbon/identity/password/expiry/util/PasswordPolicyUtils.java

```diff
@@ -250,25 +250,23 @@ private static Set<String> getUserAttributes(PasswordExpiryRuleAttributeEnum att
             throws PostAuthenticationFailedException {
 
         if (!fetchedUserAttributes.containsKey(attribute)) {
-            try {
-                switch (attribute) {
-                    case ROLES:
-                        List<RoleBasicInfo> userRoles = getUserRoles(tenantDomain, userId);
-                        Set<String> userRoleIds = userRoles.stream().map(RoleBasicInfo::getId).collect(Collectors.toSet());
-                        fetchedUserAttributes.put(PasswordExpiryRuleAttributeEnum.ROLES, userRoleIds);
-                        break;
-                    case GROUPS:
-                        List<Group> userGroups =
-                                ((AbstractUserStoreManager) userStoreManager).getGroupListOfUser(userId,
-                                        null, null);
-                        Set<String> userGroupIds = userGroups.stream().map(Group::getGroupID).collect(Collectors.toSet());
-                        fetchedUserAttributes.put(PasswordExpiryRuleAttributeEnum.GROUPS, userGroupIds);
-                        break;
-                }
-            } catch (UserStoreException e) {
-                throw new PostAuthenticationFailedException(PasswordPolicyConstants.ErrorMessages.
-                        ERROR_WHILE_RETRIEVING_USER_GROUPS.getCode(),
-                        PasswordPolicyConstants.ErrorMessages.ERROR_WHILE_RETRIEVING_USER_GROUPS.getMessage());
+            switch (attribute) {
+                case ROLES:
+                    // Fetch roles assigned to user via groups.
+                    Set<String> userGroupIds = getUserGroupIds(userId, userStoreManager);
+                    List<String> roleIdsOfGroups = getRoleIdsOfGroups(new ArrayList<>(userGroupIds), tenantDomain);
+                    fetchedUserAttributes.put(PasswordExpiryRuleAttributeEnum.GROUPS, userGroupIds);
+
+                    List<RoleBasicInfo> userRoles = getUserRoles(tenantDomain, userId);
+                    Set<String> userRoleIds =
+                            userRoles.stream().map(RoleBasicInfo::getId).collect(Collectors.toSet());
+                    userRoleIds.addAll(roleIdsOfGroups);
+                    fetchedUserAttributes.put(PasswordExpiryRuleAttributeEnum.ROLES, userRoleIds);
+                    break;
+                case GROUPS:
+                    Set<String> groupIds = getUserGroupIds(userId, userStoreManager);
```

Comment:
check fetchedUserAttributes before proceeding. As this might be done already. Have the logic to above case also.

Reply:
That's checked already before going to Switch
```
 if (!fetchedUserAttributes.containsKey(attribute)) {
 switch (attribute) {
     case ROLES:
     case GROUPS:
```

I added an additional check [1] inside Roles switch case, before retrieving userGroups to get roles assigned via groups.
[1] - https://github.com/wso2-extensions/identity-governance/pull/846/commits/8649ee5ea83b2859e5b7871078fc4b30a7e3262a